src/compute-client/src/plan/mod.rs

```diff
@@ -1887,6 +1896,46 @@ This is not expected to cause incorrect results, but could indicate a performanc
         mz_repr::explain::trace_plan(dataflow);
     }
 
+    /// Refines the plans of objects to be built as part of a single-time `dataflow` to relax
+    /// the setting of the `must_consolidate` attribute of monotonic operators, if necessary,
+    /// whenever the input is deemed to be physically monotonic.
+    #[tracing::instrument(
+        target = "optimizer",
+        level = "debug",
+        skip_all,
+        fields(path.segment = "refine_single_time_consolidation")
+    )]
+    fn refine_single_time_consolidation(
+        dataflow: &mut DataflowDescription<Self>,
+    ) -> Result<(), RecursionLimitError> {
+        // Check if we have a one-shot SELECT query, i.e., a single-time dataflow.
+        if !dataflow.is_single_time() {
+            return Ok(());
+        }
+
+        // Obtain monotonic source and index global IDs, then apply transform.
+        let monotonic_ids = dataflow
+            .source_imports
+            .iter()
+            .filter_map(|(id, (_, monotonic))| if *monotonic { Some(id) } else { None })
+            .chain(
+                dataflow
+                    .index_imports
+                    .iter()
+                    .filter_map(|(id, (_, _, monotonic))| if *monotonic { Some(id) } else { None }),
+            )
+            .cloned()
+            .collect::<BTreeSet<_>>();
+
+        let transform = transform::RelaxMustConsolidate::<T>::new();
+        let config = TransformConfig { monotonic_ids };
```

Comment:
nit: if we add more transforms this `config` instance will be shared across all of them, so it probably makes more sense to have it as a `refine_single_time_consolidation` parameter and construct it in `finalize_dataflow`.